oddt/toolkits/extras/rdkit.py

```diff
@@ -156,10 +160,315 @@ def _atom_matches_smarts(atom, smarts):
+                               % (i + 1, a1 + 1, a2 + 1))
+
+        # Fragment molecule on bonds to ge rigid fragments
+        frags = list(Chem.GetMolFrags(
+            Chem.FragmentOnBonds(mol, [mol.GetBondBetweenAtoms(a1, a2).GetIdx()
+                                       for a1, a2 in bond_atoms], addDummies=False)))
+
+        def weigh_frags(frag):
+            """sort by the fragment size and the number of bonds (secondary)"""
+            num_bonds = 0
+            # bond_weight = 0
+            for a1, a2 in bond_atoms:
+                if a1 in frag or a2 in frag:
+                    num_bonds += 1
+                    # for frag2 in frags:
+                    #     if a1 in frag2 or a2 in frag2:
+                    #         bond_weight += len(frag2)
+
+            # changed signs are fixing mixed sorting type (ascending/descending)
+            return -len(frag), -num_bonds,  # bond_weight
+        frags = sorted(frags, key=weigh_frags)
+
+        # Start writting the lines with ROOT
+        pdbqt_lines.append('ROOT')
+        frag = frags.pop(0)
+        for idx in frag:
+            pdbqt_lines.append(atom_lines[idx])
+        pdbqt_lines.append('ENDROOT')
+
+        # Now build the tree of torsions usign DFS algorithm. Keep track of last
+        # route with following variables to move down the tree and close branches
+        branch_queue = []
+        current_root = frag
+        old_roots = [frag]
+
+        visited_frags = []
+        visited_bonds = []
+        while len(frags) > len(visited_frags):
+            end_branch = True
+            for frag_num, frag in enumerate(frags):
+                for bond_num, (a1, a2) in enumerate(bond_atoms):
+                    if (frag_num not in visited_frags and
+                        bond_num not in visited_bonds and
+                        (a1 in current_root and a2 in frag or
+                         a2 in current_root and a1 in frag)):
+                        # direction of bonds is important
+                        if a1 in current_root:
+                            bond_dir = '%i %i' % (a1 + 1, a2 + 1)
+                        else:
+                            bond_dir = '%i %i' % (a2 + 1, a1 + 1)
+                        pdbqt_lines.append('BRANCH %s' % bond_dir)
+                        for idx in frag:
+                            pdbqt_lines.append(atom_lines[idx])
+                        branch_queue.append('ENDBRANCH %s' % bond_dir)
+
+                        # Overwrite current root and stash previous one in queue
+                        old_roots.append(current_root)
+                        current_root = frag
+
+                        # remove used elements from stack
+                        visited_frags.append(frag_num)
+                        visited_bonds.append(bond_num)
+
+                        # mark that we dont want to end branch yet
+                        end_branch = False
+                        break
+                    else:
```

Comment:
I think it might be confusing for some people, comment like `break outer loop as well` might help